typed/remove_test.go

```diff
@@ -0,0 +1,133 @@
+package typed
+
+import (
+	"fmt"
+	"testing"
+
+	"sigs.k8s.io/structured-merge-diff/v3/fieldpath"
+	"sigs.k8s.io/structured-merge-diff/v3/value"
+)
+
+func TestRemove(t *testing.T) {
+
+	var cases = []struct {
+		object YAMLObject
+		remove *fieldpath.Set
+		deep   bool
+		expect YAMLObject
+	}{
+		{
+			object: `{}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: ``,
+		},
+		{
+			object: `{"a": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `null`,
+		},
+		{
+			object: `{"a": "value", "b": "value"}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `{"b": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b": {}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("a")),
+			deep:   true,
+			expect: `{"b": {}}`,
+		},
+		{
+			object: `{"a": "value", "b": {"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			deep:   true,
+			expect: `{"a": "value"}`,
+		},
+		{
+			object: `{"a": "value", "b":{"c":"value"}}`,
+			remove: fieldpath.NewSet(fieldpath.MakePathOrDie("b")),
+			deep:   false,
+			// TODO: this seems to behave differently with a non deduced schema: {"a": "value", "b":{"c":"value"}} which is the reason for this PR
```

Comment:
@jennybuckley probably knows, I suspect it's because we treat deduced fields as map fields?

Reply:
I'm thinking about adding a second test that defines the schema for all the tests above to verify.
If the behavior differs, we can at least be sure we want it to do.

Reply:
Yes I think this is because deduced schema would see "b":{"c":"value"} as a map item, and so it would be removable, but if you defined a schema which has the fields named, then it would be seen as a struct field, and I don't think we expect to remove those with RemoveItems, since this function was created to remove list/map items that no appliers owns anymore, a behavior which doesn't apply to struct fields (since we didn't want removing replicas from a config and reapplying to reset replicas to 1)